src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,177 @@
+#include <sc/sidechainTxsCommitmentBuilder.h>
+#include <primitives/transaction.h>
+#include <primitives/certificate.h>
+
+SidechainTxsCommitmentBuilder::EmptyField::EmptyField():
+    zerosHelperArray{'\0'}, zero(zendoo_deserialize_field(zerosHelperArray))
+{}
+
+SidechainTxsCommitmentBuilder::EmptyField::~EmptyField()
+{
+    zendoo_field_free(const_cast<field_t*>(zero));
+}
+
+SidechainTxsCommitmentBuilder::EmptyField SidechainTxsCommitmentBuilder::emptyField;
+
+#ifdef BITCOIN_TX
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx) { return; }
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert) { return; }
+uint256 SidechainTxsCommitmentBuilder::getCommitment() { return uint256(); }
+#else
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx)
+{
+    if (!tx.IsScVersion())
+        return;
+
+    unsigned int ccOutIdx = 0;
+    LogPrint("sc", "%s():%d -getting leaves for vsc out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVscCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVscCcOut().at(ccOutIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVscCcOut()[ccOutIdx].GetHash(), tx.GetHash(), ccOutIdx);
+        vec.push_back(pField);
+    }
+
+    LogPrint("sc", "%s():%d -getting leaves for vft out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVftCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVftCcOut().at(ccOutIdx).GetScId();
```

Comment:
In order to avoid useless copy, use can be made of const uint256&.

Reply:
as discussed AFK, gcc will use return value optimization even with -O0, and will not use it only if the flag  -fno-elide-constructors is specified. If you agree, I'd avoid the reference construct.